<h1 ><center> Data Analysis of Deployed Google Assistant Actions</center></h1>
<h1 style="color:Tomato;"><center>#Hey Google</center></h1>

<h2><center>July - August 2019</center></h2>

<h2 style="color:MediumSeaGreen;"><center>Emmanuel Acheampong</center></h2>

![Google Actions](1.gif)

### After exploring cloud technology for a while, I decided to dip my toes into building actions for the Google Assistant and Below is an analysis of the 4 actions I deployed using DialogFlow, Firebase Storage and a host of Google Cloud Platform products.
<br>

I decided to use my Data Science skills to gain insights into the use of my actions and generate more value for my users.

<br>
<br>
<br>
<br>
<br>
<font size="+3"style="color:Red;"><center>ACTIONS</center></font>

<center><ul>
    <h2>Anansi Stories </h2>
    <h2>African Proverbs </h2>
    <h2>Akan Translate </h2>
    <h2>Ghana Trivia </h2>
</ul></center>

<br>
<br>
<br>

<ul><li><font size="+2" style="color:Green;">Anansi Stories</font></li></ul>

<img src="img/an.jpg" style="width:300px;height:400px;" align="left">

<h1 style="color:Green;">Data Collection</h1>

The great thing about the Google Actions Console is that it comes with some data analytics integrated with it. The console gives insight into: 

<ul>
    <li> Active Users </li>
    <li> Messages </li>
    <li> Conversations </li>
    <li> Average conversation length </li>
    <li> Abort rate </li>
</ul>

<br>
Even though all these insights are beneficial, I decided to focus mainly on "Messages", "Conversations" and "Active Users". This is because I know that these are the most relevant to improving the user experience of my users.



<h1 style="color:Green;">Conversations</h1>

In [68]:
# import modules
import pandas as pd
import numpy as np
import plotly as py
import plotly.graph_objs as go
from plotly.offline import iplot
py.offline.init_notebook_mode(connected=True) 

In [59]:
# load conversations csv and read only part of the data
anansi_conversations = pd.read_csv('data/an.csv')
anansi_conversations.head()

Date  Conversations
0  12-Jul-19              6
1  13-Jul-19             26
2  14-Jul-19             41
3  15-Jul-19             84
4  16-Jul-19             44

In [60]:
# load the other conversations csv and read only part of the data
anansi_conversations2 = pd.read_csv('data/an4.csv')
anansi_conversations2.head()

Date  Conversations
0   8-Jul-19              8
1   9-Jul-19             42
2  10-Jul-19            133
3  11-Jul-19             30
4  12-Jul-19              6

In [138]:
# merge the 2 csv and remove duplicate files
anansi_convos = pd.concat([anansi_conversations2,anansi_conversations]).drop_duplicates(subset='Date', keep="last")

<br>
<br>
<br>
<h1 ><center> Visualization </center></h1>

In [81]:
colors = np.random.rand(100)
sz = np.random.rand(100)*30
fig = go.Figure()
fig.add_scatter(x = anansi_convos['Date'],
                y = anansi_convos['Conversations'],
                mode='markers',
                marker={'size': 10,
                        'color': 'Red',
                        'opacity': 0.6,
                        'colorscale': 'Viridis'
                       });

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number of Conversations')

fig.show()

In [80]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=anansi_convos['Date'],
    y=anansi_convos['Conversations'],
    name = '<b>Number of Daily Conversations</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
))

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number of Conversations')

fig.show()

<br>
<br>
<br>
<br>

# Data Analysis

In [139]:
anansi_convos['Conversations'].agg(['sum', 'min', 'max',"mean", 'std'])

sum     1139.000000
min        3.000000
max      133.000000
mean      34.515152
std       37.022899
Name: Conversations, dtype: float64

### The Number of Conversations had in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>1139</center></font>

<br>
<br>

### The Minimum Daily Conversations had in One Month (from July 8th to August 9th): 

<font size="+3"style="color:Red;"><center>3</center></font>

<br>
<br>

### The Maximum Daily Conversations had in One Month (from July 8th to August 9th): 

<font size="+3"style="color:Red;"><center>133</center></font>

<br>
<br>

### The Average Daily Conversations had in One Month (from July 8th to August 9th): 

<font size="+3"style="color:Red;"><center>34</center></font>

<br>
<br>

### The Standard Deviation of Daily Conversations had in One Month (from July 8th to August 9th): 

<font size="+3"style="color:Red;"><center>37</center></font>

<br>
<br>
<br>
<br>
<br>
<br>

<h1 style="color:Green;">Messages</h1>

In [100]:
# load messages 
anansi_messages = pd.read_csv('data/an3.csv')
anansi_messages.head()

Date  Messages
0   8-Jul-19         8
1   9-Jul-19        42
2  10-Jul-19       133
3  11-Jul-19        30
4  12-Jul-19         6

In [101]:
# load messages 
anansi_messages2 = pd.read_csv('data/an7.csv')
anansi_messages2.head()

Date  Messages
0  12-Jul-19         6
1  13-Jul-19        26
2  14-Jul-19        41
3  15-Jul-19        84
4  16-Jul-19        44

In [113]:
# merge the 2 csv and remove duplicate files
anansi_msg = pd.concat([anansi_messages,anansi_messages2]).drop_duplicates(subset='Date', keep="last")

<br>
<br>
<br>
<h1 ><center> Visualization </center></h1>

In [107]:
colors = np.random.rand(100)
sz = np.random.rand(100)*30
fig = go.Figure()
fig.add_scatter(x = anansi_msg['Date'],
                y = anansi_msg['Messages'],
                mode='markers',
                marker={'size': 10,
                        'color': 'Red',
                        'opacity': 0.6,
                        'colorscale': 'Viridis'
                       });

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number of Messages')

fig.show()

In [109]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=anansi_msg['Date'],
    y=anansi_msg['Messages'],
    name = '<b>Number of Daily Conversations</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
))

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number of Messages')

fig.show()

<br>
<br>
<br>
<br>

# Data Analysis

In [115]:
anansi_msg['Messages'].agg(['sum', 'min', 'max',"mean", 'std'])

sum     1139.000000
min        3.000000
max      133.000000
mean      34.515152
std       37.022899
Name: Messages, dtype: float64

<br>
<br>
<br>
<br>

### The Number of Messages had in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>1139</center></font>

<br>
<br>


### The Minimum Daily Messages had in One Month (from July 8th to August 9th): 



<font size="+3"style="color:Red;"><center>3</center></font>



<br>
<br>

### The Maximum Daily Messages had in One Month (from July 8th to August 9th): 



<font size="+3"style="color:Red;"><center>133</center></font>

<br>
<br>


### The Average Daily Messages had in One Month (from July 8th to August 9th): 





<font size="+3"style="color:Red;"><center>34</center></font>

<br>
<br>

### The Standard Deviation of Daily Messages had in One Month (from July 8th to August 9th): 




<font size="+3"style="color:Red;"><center>37</center></font>

<br>
<br>

<br>
<br>
<br>
<h1 style="color:Green;">Active Users</h1>

In [131]:
# load messages 
anansi_users = pd.read_csv('data/an8.csv')
anansi_users.head()

Date  Total
0  12-Jul-19      6
1  13-Jul-19     26
2  14-Jul-19     39
3  15-Jul-19     80
4  16-Jul-19     42

In [132]:
# load messages 
anansi_users2 = pd.read_csv('data/an2.csv')
anansi_users2.head()

Date  Total
0   8-Jul-19      2
1   9-Jul-19     40
2  10-Jul-19    119
3  11-Jul-19     28
4  12-Jul-19      6

In [135]:
# merge the 2 csv and remove duplicate files
anansi_users = pd.concat([anansi_users2,anansi_users]).drop_duplicates(subset='Date', keep="last")

<br>
<br>
<br>
<h1 ><center> Visualization </center></h1>

In [140]:
colors = np.random.rand(100)
sz = np.random.rand(100)*30
fig = go.Figure()
fig.add_scatter(x = anansi_users['Date'],
                y = anansi_users['Total'],
                mode='markers',
                marker={'size': 10,
                        'color': 'Red',
                        'opacity': 0.6,
                        'colorscale': 'Viridis'
                       });

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number of Users')

fig.show()

In [141]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=anansi_users['Date'],
    y=anansi_users['Total'],
    name = '<b>Number of Daily Users</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
))

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number of Users')

fig.show()

<br>
<br>
<br>
<br>

# Data Analysis

In [142]:
anansi_users['Total'].agg(['sum', 'min', 'max',"mean", 'std'])

sum     1008.000000
min        2.000000
max      119.000000
mean      30.545455
std       33.189881
Name: Total, dtype: float64

<br>
<br>
<br>
<br>

### The Number of Active Users in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>1008</center></font>

### The Minimum Daily Users had in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>2</center></font>

### The Maximum Daily Users had in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>119</center></font>

### The Average number of Daily Users had in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>30</center></font>

### The Standard Deviation of Daily Users had in One Month (from July 8th to August 9th): 

<font size="+3"style="color:Red;"><center>33</center></font>

<br>
<br>
<br>
<br>
<br>
<br>
## Merging all the Active Users and Conversations

In [184]:
finaldf = pd.concat([anansi_users, anansi_msg, anansi_convos], axis=1, join='inner').sort_index()

In [289]:
finaldf.columns = ['Date', 'Anansi Stories Active Users', 'Anansi Stories Conversations']

Date  Active Users  Conversations
0   8-Jul-19             2              8
0  12-Jul-19             6              6
1   9-Jul-19            40             42
1  13-Jul-19            26             26
2  10-Jul-19           119            133

In [197]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=finaldf['Date'],
    y=finaldf['Active Users'],
    name = '<b>Number of Active Users of Anansi Stories</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
))

fig.add_trace(go.Scatter(
    x=finaldf['Date'],
    y=finaldf['Conversations'],
    name = '<b>Number of Conversations on Anansi Stories</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
))

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number')

fig.show()

In [264]:

fig = go.Figure()
fig.add_trace(go.Scatter(x=finaldf['Date'],y=finaldf['Conversations'], fill='tozeroy', name = '<b>Number of Active Users of Anansi Stories</b>')) # fill down to xaxis
fig.add_trace(go.Scatter(x=finaldf['Date'],y=finaldf['Active Users'], fill='tonexty', name = '<b>Number of Conversations on Anansi Stories</b>')) # fill to trace0 y

fig.show()

<br>
<br>
<br>
<br>
<br>
<br>

<ul><li><font size="+2" style="color:Green;">African Proverbs</font></li></ul>

<img src="img/ap.jpg" style="width:300px;height:400px;" align="left">

<h1 style="color:Green;">Conversations</h1>

In [231]:
ap_convos = pd.read_csv('data/ap7.csv')

In [232]:
ap_convos.head()

Date1  Conversations
0  11-Jul-19              3
1  13-Jul-19              2
2  15-Jul-19             18
3  16-Jul-19              7
4  17-Jul-19             11

<h1 style="color:Green;">Active Users</h1>

In [233]:
ap_convos2 = pd.read_csv('data/ap8.csv')

In [234]:
ap_convos2.head()

Date  Total
0  11-Jul-19      3
1  12-Jul-19      0
2  13-Jul-19      2
3  14-Jul-19      0
4  15-Jul-19      6

In [282]:
finaldf2 = pd.concat([ap_convos2, ap_convos], axis=1, join='inner').sort_index()

In [283]:
del finaldf2['Date1']

In [284]:
finaldf2.columns = ['Date', 'Active Users','Conversations']

In [285]:
finaldf2.tail()

Date  Active Users  Conversations
23  3-Aug-19            14             19
24  4-Aug-19            22             32
25  5-Aug-19            18             39
26  6-Aug-19            23             19
27  7-Aug-19            17             19

In [263]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=finaldf2['Date'],
    y=finaldf2['Active Users'],
    line_color='red',
    name = '<b>Number of Active Users of African Proverbs</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
    
))

fig.add_trace(go.Scatter(
    x=finaldf2['Date'],
    y=finaldf2['Conversations'],
    line_color='green',
    name = '<b>Number of Conversations on African Proverbs</b>', # Style name/legend entry with html tags
    connectgaps=True # override default to connect the gaps
))

fig.layout.update(title='',
                   xaxis_title='Day',
                   yaxis_title='Number')

fig.show()

In [262]:

fig = go.Figure()
fig.add_trace(go.Scatter(x=finaldf2['Date'],y=finaldf2['Conversations'], fill='tozeroy',line_color='red', name = '<b>Number of Active Users of African Proverbs</b>')) # fill down to xaxis
fig.add_trace(go.Scatter(x=finaldf2['Date'],y=finaldf2['Active Users'], fill='tonexty',line_color='green',name = '<b>Number of Conversations on African Proverbs</b>')) # fill to trace0 y

fig.show()

<br>
<br>
<br>
<br>

# Data Analysis

In [266]:
finaldf2['Active Users'].agg(['sum', 'min', 'max',"mean", 'std'])

sum     275.000000
min       0.000000
max      28.000000
mean      9.821429
std       8.948634
Name: Active Users, dtype: float64

In [267]:
finaldf2['Conversations'].agg(['sum', 'min', 'max',"mean", 'std'])

sum     417.000000
min       1.000000
max      39.000000
mean     14.892857
std      11.461172
Name: Conversations, dtype: float64

In [287]:
finaldf2.columns = ['Date', 'African Proverb Active Users', 'African Proverbs Conversations']

<br>
<br>
<br>
<br>

### The Number of Active Users in One Month (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>275</center></font>

### The Minimum Number of Daily Active Users (from July 8th to August 9th): 


<font size="+3"style="color:Red;"><center>0</center></font>


### The Maximum Number of Daily Active Users (from July 8th to August 9th): 



<font size="+3"style="color:Red;"><center>28</center></font>


### The Average Number of Daily Active Users (from July 8th to August 9th): 



<font size="+3"style="color:Red;"><center>9</center></font>


### The Standard Deviation of the Number of Daily Active Users (from July 8th to August 9th): 



<font size="+3"style="color:Red;"><center>8</center></font>

<br>
<br>
<br>
<br>
<br>

### The Total Number of Conversations in One Month : 


<font size="+3"style="color:Red;"><center>417</center></font>

### The Minimum Number of Daily Conversations: 


<font size="+3"style="color:Red;"><center>1</center></font>

### The Maximum Number of Daily Conversations: 


<font size="+3"style="color:Red;"><center>39</center></font>


### The Average Number of Daily Conversations: 



<font size="+3"style="color:Red;"><center>14</center></font>

### The Standard Deviation of the Number of Daily Conversations: 


<font size="+3"style="color:Red;"><center>11</center></font>

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

In [290]:
merged = finaldf.merge(finaldf2, on='Date')

In [291]:
merged

Date  Anansi Stories Active Users  Anansi Stories Conversations  \
0   12-Jul-19                            6                             6   
1   13-Jul-19                           26                            26   
2   14-Jul-19                           39                            41   
3   11-Jul-19                           28                            30   
4   15-Jul-19                           80                            84   
5   16-Jul-19                           42                            44   
6   17-Jul-19                           69                            83   
7   18-Jul-19                           83                            89   
8   19-Jul-19                           91                           101   
9   20-Jul-19                           53                            55   
10  21-Jul-19                           47                            49   
11  22-Jul-19                          102                           124   
12  23-Jul-19                           29                            36   
13  24-Jul-19                           45                            53   
14  25-Jul-19                           14                            17   
15  26-Jul-19                            9                            12   
16  27-Jul-19                           12                            15   
17  28-Jul-19                            5                             5   
18  29-Jul-19                            8                             9   
19  30-Jul-19                            4                             4   
20  31-Jul-19                            7                             7   
21   1-Aug-19                            7                             7   
22   2-Aug-19                           10                            14   
23   3-Aug-19                            5                             5   
24   4-Aug-19                            4                            11   
25   5-Aug-19                            5                             5   
26   6-Aug-19                            4                             4   
27   7-Aug-19                            4                             4   

    African Proverb Active Users  African Proverbs Conversations  
0                              0                               2  
1                              2                              18  
2                              0                               7  
3                              3                               3  
4                              6                              11  
5                              6                               5  
6                              5                               1  
7                              4                               2  
8                              1                               2  
9                              2                              13  
10                             1                               2  
11                             4                               3  
12                             2                               4  
13                             3                               9  
14                             3                              29  
15                             3                              11  
16                            28                              20  
17                            10                              21  
18                            16                              19  
19                            19                              24  
20                            16                              30  
21                            23                              15  
22                            24                              38  
23                            14                              19  
24                            22                              32  
25                            18                              

In [297]:
fig = go.Figure()

# Add traces
fig.add_trace(go.Scatter(x=merged['Date'], y=merged['Anansi Stories Active Users'],
                    mode='lines+markers',
                    name='Anansi Stories Active Users'))
fig.add_trace(go.Scatter(x=merged['Date'], y=merged['Anansi Stories Conversations'],
                    mode='lines+markers',
                    name='Anansi Stories Conversations'))
fig.add_trace(go.Scatter(x=merged['Date'], y=merged['African Proverb Active Users'],
                    mode='lines+markers',
                    name='African Proverb Active Users'))

fig.add_trace(go.Scatter(x=merged['Date'], y=merged['African Proverbs Conversations'],
                    mode='lines+markers',
                    name='African Proverbs Conversations'))

fig.show()

<br>
<br>
<br>
<br>
<br>
<br>